File: src/main/java/WavefrontObj.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class WavefrontObj
{
    private final List<Vertex> vertices;
    private final List<Polygon> polygons;

    public WavefrontObj(List<Vertex> vertices, List<Polygon> polygons)
    {
        this.vertices = vertices;
        this.polygons = polygons;
    }

    public List<Vertex> getVertices()
    {
        return vertices;
    }

    public List<Polygon> getPolygons()
    {
        return polygons;
    }

    public static WavefrontObj parse(String fileName)
    {
        BufferedReader file;
        try
        {
            file = new BufferedReader(new FileReader(fileName));
        }
        catch (IOException e)
        {
            throw new IllegalStateException(e);
        }

        List<List<String>> splitFileLines = file.lines()
                .map(line -> Arrays.asList(line.split(" +")))
                .collect(Collectors.toList());

        List<Vertex> vertices = splitFileLines.stream()
                .filter(l -> l.get(0).equals("v"))
                .map(l -> new Vertex(Double.valueOf(l.get(1)), Double.valueOf(l.get(2)), Double.valueOf(l.get(3))))
                .collect(Collectors.toList());

        List<Polygon> polygons = splitFileLines.stream()
                .filter(l -> l.get(0).equals("f"))
                .map(l -> l.stream()
                        .skip(1)
                        .map(s -> Arrays.asList(s.split("/")).get(0))
                        .map(Integer::valueOf)
                        .map(v -> v - 1)
                        .collect(Collectors.toList()))
                .map(Polygon::new)
                .collect(Collectors.toList());

        return new WavefrontObj(vertices, polygons);
    }

    public static class Polygon
    {
        private final List<Integer> polygon;

        public Polygon(List<Integer> polygon)
        {
            this.polygon = polygon;
        }

        public Integer size()
        {
            return polygon.size();
        }

        public Integer get(int node)
        {
            return polygon.get(node % polygon.size());
        }

        public List<Integer> getPolygon()
        {
            return polygon;
        }
    }
}

File: src/main/kotlin/utils/WavefrontObj.kt
package utils

import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

class WavefrontObj(
    val vertices: List<Vec3d>,
    val textureCoords: List<Vec2d>,
    val polygons: List<Polygon>,
    val normalMap: List<Vec3d>
) {

    class Polygon(val polygon: List<Vertex>) {
        val size: Int get() = polygon.size

        operator fun get(node: Int): Vertex {
            return polygon[node % polygon.size]
        }
    }

    data class Vertex(val vertice: Int, val texture: Int)

    companion object {
        fun parse(fileName: String): WavefrontObj {
            val file: BufferedReader
            try {
                file = BufferedReader(FileReader(fileName))
            } catch (e: IOException) {
                throw IllegalStateException(e)
            }

            val splitFileLines = file.lines()
                .map { line -> line.split("\\s+".toRegex()).dropLastWhile { it.isEmpty() } }
                .filter { it.isNotEmpty() }
                .toList()

            val vertices = splitFileLines
                .filter { l -> l[0] == "v" }
                .map { l -> Vec3d(l[1].toDouble(), l[2].toDouble(), l[3].toDouble()) }
                .toList()

            val textureCoord = splitFileLines
                .filter { l -> l[0] == "vt" }
                .map { l -> Vec2d(l[1].toDouble(), l[2].toDouble()) }
                .toList()

            val normalMap = splitFileLines
                .filter { l -> l[0] == "vn" }
                .map { l -> Vec3d(l[1].toDouble(), l[2].toDouble(), l[3].toDouble()) }
                .toList()

            val polygons = splitFileLines
                .filter { l -> l[0] == "f" }
                .map { l ->
                    l.drop(1)
                        .map { s ->
                            val v = s.split("/".toRegex()).dropLastWhile { it.isEmpty() }
                            Vertex(v[0].toInt() - 1, (v.getOrNull(1)?.toInt() ?: 1) - 1)
                        }
                        .toList()
                }
                .map { Polygon(it) }
                .toList()

            return WavefrontObj(vertices, textureCoord, polygons, normalMap)
        }
    }
}